<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bears: building expression atlases in R • bears</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="bears: building expression atlases in R">
<meta property="og:description" content="bears">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bears</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/bears.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/almeidasilvaf/bears/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>bears: building expression atlases in R</h1>
                        <h4 data-toc-skip class="author">Fabricio Almeida-Silva</h4>
            <address class="author_afil">
      Universidade Estadual do Norte Fluminense Darcy Ribeiro, RJ, Brazil<br><h4 data-toc-skip class="author">Thiago Motta Venancio</h4>
            <address class="author_afil">
      Universidade Estadual do Norte Fluminense Darcy Ribeiro, RJ, Brazil<br><h4 data-toc-skip class="date">2022-05-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/almeidasilvaf/bears/blob/HEAD/vignettes/bears.Rmd" class="external-link"><code>vignettes/bears.Rmd</code></a></small>
      <div class="hidden name"><code>bears.Rmd</code></div>

    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In the past decades, there has been an exponential accumulation of RNA-seq data in public repositories. This steep increase paved the way to the creation of gene expression atlases, which consist in comprehensive collections of expression data from public databases, analyzed with a single pipeline for consistency and across-project comparison. <code>bears</code> is a package that allows you to create your own gene expression atlas for a given species using public data. The package features:</p>
<ul>
<li>Data download from NCBI’s Sequence Read Archive (SRA) or the European Nucleotide Archive (ENA).</li>
<li>Sequence quality check and trimming of low-quality sequences.</li>
<li>Removal of rRNA, which are typical problems of libraries that were prepared with the rRNA depletion protocol.</li>
<li>Read mapping against a reference genome.</li>
<li>Transcript assembly.</li>
<li>Quantification of gene and transcript expression with either <strong>alignment-based</strong> or <strong>alignment-free</strong> methods.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation-and-setup">Installation and setup<a class="anchor" aria-label="anchor" href="#installation-and-setup"></a>
</h2>
<p>To install <code>bears</code>, use the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"almeidasilvaf/bears"</span><span class="op">)</span></code></pre></div>
<p>Then, create a standard directory structure with <code><a href="../reference/create_dir_structure.html">create_dir_structure()</a></code> to store your results. This is optional, but <strong>it will make your life much easier.</strong> The output is a list of paths to common directories that you will need to specify in several functions of this package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create directory structure using a temporary directory as root</span>
<span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_dir_structure.html">create_dir_structure</a></span><span class="op">(</span>rootdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Look at the output</span>
<span class="va">ds</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="retrieve-sample-metadata">Retrieve sample metadata<a class="anchor" aria-label="anchor" href="#retrieve-sample-metadata"></a>
</h2>
<p>First of all, you need to choose which samples you want to download and create a <strong>metadata data frame</strong> for your samples. To create this data frame, you will pass a search term to the function <code><a href="../reference/create_sample_info.html">create_sample_info()</a></code>. The search term has the same syntax of the SRA search term syntax. For example, you can search by:</p>
<ul>
<li>BioSample accession - <strong>SAMN08903403[BSPL]</strong>
</li>
<li>BioProject accession - <strong>PRJNA229998[GPRJ]</strong>
</li>
<li>All BioSamples for an organism - <strong>Glycine max[ORGN] AND RNA-seq[STRA]</strong>
</li>
<li>And so on…</li>
</ul>
<p>Let’s create a metadata data frame for a human RNA-seq sample that is included in the {airway} Bioconductor package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create metadata data frame</span>
<span class="va">term</span> <span class="op">&lt;-</span> <span class="st">"SAMN02422669[BSPL]"</span>
<span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_sample_info.html">create_sample_info</a></span><span class="op">(</span><span class="va">term</span><span class="op">)</span>
<span class="va">metadata</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="download-fastq-files">Download FASTQ files<a class="anchor" aria-label="anchor" href="#download-fastq-files"></a>
</h2>
<p>To download the .fastq files with the reads, you have 2 options:</p>
<ul>
<li>Using <code><a href="../reference/download_fastq.html">download_fastq()</a></code>, which relies on the popular SRA Toolkit to download reads from NCBI’s SRA. You need to have SRA Toolkit installed in your machine to run this function.</li>
<li>Using <code><a href="../reference/download_from_ena.html">download_from_ena()</a></code> (recommended), which downloads reads from ENA. This function does not require any external dependency and it is much faster than using SRA Toolkit.</li>
</ul>
<p>As input, you only need to give the metadata data frame and the path the output directory where .fastq files will be stored. For example:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Download sample to temporary directory</span>
<span class="va">download</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_from_ena.html">download_from_ena</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sequence-quality-check-and-trimming">Sequence quality check and trimming<a class="anchor" aria-label="anchor" href="#sequence-quality-check-and-trimming"></a>
</h2>
<p>After downloading the .fastq files, you need to perform some quality checks to check if you don’t have poor-quality base calling or sequence adapters. These can be done with the function <code><a href="../reference/run_fastqc.html">run_fastqc()</a></code>, which (guess what?) runs FastQC from the R session. FastQC results can then be summarized with MultiQC.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">fastqc_table</span><span class="op">)</span>

<span class="co"># Run FastQC for all .fastq files in fastqdir</span>
<span class="va">multiqc_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"multiqc"</span><span class="op">)</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="../reference/fastqc_is_installed.html">fastqc_is_installed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="../reference/multiqc_is_installed.html">multiqc_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">fastqc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_fastqc.html">run_fastqc</a></span><span class="op">(</span>
        <span class="va">metadata</span>, 
        fastqdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span>, 
        fastqcdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqcdir</span>
    <span class="op">)</span>
    <span class="va">fastqc_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multiqc.html">multiqc</a></span><span class="op">(</span><span class="va">ds</span><span class="op">$</span><span class="va">fastqcdir</span>, <span class="va">multiqc_out</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># Look at the summary output</span>
<span class="va">fastqc_table</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-filtering">Read filtering<a class="anchor" aria-label="anchor" href="#read-filtering"></a>
</h2>
<p>Read filtering consists in 2 steps:</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/trim_reads.html">trim_reads()</a></code> - trim adapters and low-quality bases. This function runs Trimmomatic from an R session and saves filtered .fastq files in a directory named <code>filtdir</code>.</li>
<li>
<code><a href="../reference/remove_rrna.html">remove_rrna()</a></code> - remove rRNA from .fastq files, if there are any. rRNA removal relies on the SortMeRNA program.</li>
</ol>
<p>Here, as you can see in the <em>fastqc_table</em> above, the files passed the “Per Base Sequence Quality” check, so we don’t need to trim anything. For this particular case, we can skip <code><a href="../reference/trim_reads.html">trim_reads()</a></code> and only run <code><a href="../reference/remove_rrna.html">remove_rrna()</a></code>. As rRNA database, we will use an example file in this package’s /extdata subdirectory.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Specify path to rRNA db</span>
<span class="va">rrna_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"bac_16s_subset.fa"</span>, package<span class="op">=</span><span class="st">"bears"</span><span class="op">)</span>

<span class="co"># Run SortMeRNA</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="../reference/sortmerna_is_installed.html">sortmerna_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="../reference/remove_rrna.html">remove_rrna</a></span><span class="op">(</span>
        <span class="va">metadata</span>, 
        fastqdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span>,
        filtdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">filtdir</span>
    <span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>Now that we have performed all quality checks, we’re good to go.</p>
</div>
<div class="section level2">
<h2 id="read-mapping-optional">Read mapping (optional)<a class="anchor" aria-label="anchor" href="#read-mapping-optional"></a>
</h2>
<p>Read mapping is only required if you want to quantify the expression with alignment-based methods (e.g., StringTie and featureCounts) or if you want to assemble transcripts. In <code>bears</code>, reads are mapped to the reference genome with STAR.</p>
<p>Here, for the purpose of demonstration, we will map reads to a subset of the human genome. The FASTA and GTF files corresponding to the subset of the genome are available in /extdata.</p>
<p>Before mapping reads, we need to create a genome index. This can be done with <code><a href="../reference/star_genome_index.html">star_genome_index()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get paths to genome subset</span>
<span class="va">genome_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Hsapiens_GRCh37.75_subset.fa"</span>, 
                            package<span class="op">=</span><span class="st">"bears"</span><span class="op">)</span>
<span class="va">gff_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh37.75_subset.gtf"</span>, 
                         package<span class="op">=</span><span class="st">"bears"</span><span class="op">)</span>

<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="../reference/star_is_installed.html">star_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="../reference/star_genome_index.html">star_genome_index</a></span><span class="op">(</span><span class="va">genome_path</span>, <span class="va">gff_path</span>, <span class="va">ds</span><span class="op">$</span><span class="va">mapping_dir</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>Now that we have the genome index, we can map reads to it.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="../reference/star_is_installed.html">star_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="../reference/star_align.html">star_align</a></span><span class="op">(</span>
        <span class="va">metadata</span>, 
        filtdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">filtdir</span>, 
        fastqc_table <span class="op">=</span> <span class="va">fastqc_table</span>, 
        mappingdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span>,
        gff_path <span class="op">=</span> <span class="va">gff_path</span>
    <span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>Finally, let’s get mapping stats.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="../reference/multiqc_is_installed.html">multiqc_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">star_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multiqc.html">multiqc</a></span><span class="op">(</span>dir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span>,
                          outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,
                          runon <span class="op">=</span> <span class="st">"star"</span><span class="op">)</span>
    
    <span class="co"># Check if samples passed the filtering criterion</span>
    <span class="va">align_passed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapping_pass.html">mapping_pass</a></span><span class="op">(</span><span class="va">star_stats</span>, <span class="va">sample_info</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quantification">Quantification<a class="anchor" aria-label="anchor" href="#quantification"></a>
</h2>
<p>Before quantification, we need to infer library strandedness.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Convert GFF to BED</span>
<span class="fu"><a href="../reference/gff2bed.html">gff2bed</a></span><span class="op">(</span>gffpath <span class="op">=</span> <span class="st">"data/Gmax_a2.v1.gff3"</span><span class="op">)</span>

<span class="co"># Infer strandedness</span>
<span class="va">sample_info2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_strandedness.html">infer_strandedness</a></span><span class="op">(</span>
  mapping_passed <span class="op">=</span> <span class="va">align_passed</span>,
  bedpath <span class="op">=</span> <span class="st">"data/Gmax_a2.v1.bed"</span>,
  mappingdir<span class="op">=</span><span class="st">"results/04_read_mapping"</span>
  <span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Fabrício Almeida-Silva, Thiago Venancio.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
