<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bears">
<title>bears: building expression atlases in R • bears</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.2/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="bears: building expression atlases in R">
<meta property="og:description" content="bears">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#0092ac"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bears</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bears_vignette.html">bears: building expression atlases in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/almeidasilvaf/bears/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>bears: building expression atlases in R</h1>
                        <h4 data-toc-skip class="author">Fabricio
Almeida-Silva</h4>
            <address class="author_afil">
      Universidade Estadual do Norte Fluminense Darcy Ribeiro, RJ,
Brazil<br><h4 data-toc-skip class="author">Thiago
Motta Venancio</h4>
            <address class="author_afil">
      Universidade Estadual do Norte Fluminense Darcy Ribeiro, RJ,
Brazil<br><h4 data-toc-skip class="date">2022-09-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/almeidasilvaf/bears/blob/HEAD/vignettes/bears_vignette.Rmd" class="external-link"><code>vignettes/bears_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>bears_vignette.Rmd</code></div>
    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In the past decades, there has been an exponential accumulation of
RNA-seq data in public repositories. This steep increase paved the way
to the creation of gene expression atlases, which consist in
comprehensive collections of expression data from public databases,
analyzed with a single pipeline for consistency and cross-project
comparison. <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em> is a
package that allows you to create your own gene expression atlas for a
given species using public data. The package features:</p>
<ul>
<li>Data download from NCBI’s Sequence Read Archive (SRA) or the
European Nucleotide Archive (ENA).</li>
<li>Sequence quality check and trimming of low-quality sequences.</li>
<li>Removal of rRNA, which are typical problems of libraries that were
prepared with the rRNA depletion protocol.</li>
<li>Read mapping against a reference genome.</li>
<li>Transcript assembly.</li>
<li>Quantification of gene- and transcript-level transcript abundance
with <strong>alignment-based</strong> and
<strong>alignment-free</strong> methods.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation-and-setup">Installation and setup<a class="anchor" aria-label="anchor" href="#installation-and-setup"></a>
</h2>
<p>To install <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em>, use the
following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"almeidasilvaf/bears"</span><span class="op">)</span></span></code></pre></div>
<p>Then, create a standard directory structure with
<code><a href="../reference/create_dir_structure.html">create_dir_structure()</a></code> to store your results. This is
optional, but <strong>it will make your life much easier.</strong> The
output is a list of paths to common directories that you will need to
specify in several functions of this package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create directory structure using a temporary directory as root</span></span>
<span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_dir_structure.html">create_dir_structure</a></span><span class="op">(</span>rootdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at the output</span></span>
<span><span class="va">ds</span></span></code></pre></div>
<p>To run the full pipeline implemented in <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em>, you will
need to have some external software tools installed in your machine. The
names of the tools are listed below. In the <strong>Functions</strong>
column, you can see the names of functions in <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em> that
require each tool.</p>
<table class="table">
<thead><tr class="header">
<th align="left">Software</th>
<th align="left">Version</th>
<th align="left">Functions</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">FastQC</td>
<td align="left">&gt;=0.11.9</td>
<td align="left">run_fastqc()</td>
</tr>
<tr class="even">
<td align="left">kallisto</td>
<td align="left">&gt;=0.11.9</td>
<td align="left">kallisto_index(), kallisto_quantify()</td>
</tr>
<tr class="odd">
<td align="left">MultiQC</td>
<td align="left">&gt;=1.12</td>
<td align="left">multiqc()</td>
</tr>
<tr class="even">
<td align="left">RSeQC</td>
<td align="left">&gt;=4.0.0</td>
<td align="left">infer_strandedness()</td>
</tr>
<tr class="odd">
<td align="left">salmon</td>
<td align="left">&gt;=1.8.0</td>
<td align="left">salmon_index(), salmon_quantify()</td>
</tr>
<tr class="even">
<td align="left">SortMeRNA</td>
<td align="left">&gt;=4.3.4</td>
<td align="left">remove_rrna()</td>
</tr>
<tr class="odd">
<td align="left">STAR</td>
<td align="left">&gt;=2.7.10a</td>
<td align="left">star_genome_index(), star_align()</td>
</tr>
<tr class="even">
<td align="left">StringTie</td>
<td align="left">&gt;=2.2.1</td>
<td align="left">stringtie_assemble(), stringtie_quantify()</td>
</tr>
<tr class="odd">
<td align="left">Subread</td>
<td align="left">&gt;=2.0.1</td>
<td align="left">fcount()</td>
</tr>
<tr class="even">
<td align="left">TACO</td>
<td align="left">&gt;=0.7.3</td>
<td align="left">taco_merge()</td>
</tr>
<tr class="odd">
<td align="left">Trimmomatic</td>
<td align="left">&gt;=0.39</td>
<td align="left">trim_reads()</td>
</tr>
</tbody>
</table>
<p>To make your life easier, we have created .yml files with Conda
environments containing each of these external tools. In this vignette,
we will manage external dependencies with the Bioconductor package
<em><a href="https://bioconductor.org/packages/3.15/Herper" class="external-link">Herper</a></em>. We
will use <code>BiocStyle::Biocpkg("Herper")</code> to:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Install miniconda</strong>. For this vignette, we will
install miniconda in a temporary directory, which means that miniconda
will be automatically deleted when the R session closes.</li>
</ol>
<blockquote>
<p>“But I already have miniconda installed!”</p>
</blockquote>
<p>Great! In that case, just create an object (e.g.,
<code>my_miniconda</code>) with the path to your miniconda
installation.</p>
<blockquote>
<p>“Can I have a permanent installation of miniconda instead of
installing it in a temporary directory?”</p>
</blockquote>
<p>For sure. You can specify a path to a different directory in your
object containing the miniconda installation path. For instance, you can
try <code>my_miniconda &lt;- "~/Documents/miniconda"</code>.</p>
<ol start="2" style="list-style-type: decimal">
<li><p><strong>Create a different environment for each external
tool</strong>. To avoid conflicts, it is important to keep each tool in
its own environment. Here, each tool will be installed in an environment
named <code>&lt;tool-name-lowercase&gt;_env</code> (e.g.,
<code>star_env</code>, <code>rseqc_env</code>,
<code>fastqc_env</code>).</p></li>
<li><p><strong>Run R functions that require external tools</strong>. We
will run functions that call external tools inside a call to
<code>Herper::withCondaEnv()</code>. This <em><a href="https://bioconductor.org/packages/3.15/Herper" class="external-link">Herper</a></em>
function allows us to run an R function inside a particular Conda
environment, which we need to specify.</p></li>
</ol>
<p>Below is the code to install miniconda and create tool-specific
environments from the .yml files.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RockefellerUniversity/Herper" class="external-link">Herper</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Path to directory where miniconda will be installed</span></span>
<span><span class="va">my_miniconda</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"miniconda"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Path to .yml files to create environments</span></span>
<span><span class="va">envs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"bears"</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">".yml"</span>,</span>
<span>    full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install miniconda in `my_miniconda` and create envs</span></span>
<span><span class="va">create_envs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">envs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/Herper/man/SaveEnvironments.html" class="external-link">import_CondaEnv</a></span><span class="op">(</span><span class="va">x</span>, pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Let’s see the environments and their names.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/ListCondaEnvironments.html" class="external-link">list_CondaEnv</a></span><span class="op">(</span>pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span></span>
<span>    PATH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"PATH"</span><span class="op">)</span>,</span>
<span>        <span class="st">"/opt/FastQC/fastqc"</span>,</span>
<span>        <span class="st">"/opt/STAR-2.7.9a/bin/Linux_x86_64_static/"</span>,</span>
<span>        <span class="st">"/opt/bin/"</span>,</span>
<span>        <span class="st">"/opt/salmon-1.5.2_linux_x86_64/bin/"</span>,</span>
<span>        <span class="st">"/opt/kallisto/"</span>,</span>
<span>        <span class="st">"/opt/subread-2.0.3-Linux-x86_64/bin/"</span>,</span>
<span>        <span class="st">"/opt/stringtie-2.1.7.Linux_x86_64/"</span>,</span>
<span>        <span class="st">"/opt/taco-v0.7.3.Linux_x86_64/"</span>,</span>
<span>        sep <span class="op">=</span> <span class="st">":"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"PATH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/FastQC/fastqc:/opt/STAR-2.7.9a/bin/Linux_x86_64_static/:/opt/bin/:/opt/salmon-1.5.2_linux_x86_64/bin/:/opt/kallisto/:/opt/subread-2.0.3-Linux-x86_64/bin/:/opt/stringtie-2.1.7.Linux_x86_64/:/opt/taco-v0.7.3.Linux_x86_64/"</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Test software installation</span></span>
<span><span class="fu"><a href="../reference/fastqc_is_installed.html">fastqc_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="fu"><a href="../reference/multiqc_is_installed.html">multiqc_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/star_is_installed.html">star_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="fu"><a href="../reference/trimmomatic_is_installed.html">trimmomatic_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="fu"><a href="../reference/sortmerna_is_installed.html">sortmerna_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/rseqc_is_installed.html">rseqc_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="fu"><a href="../reference/salmon_is_installed.html">salmon_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/kallisto_is_installed.html">kallisto_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/subread_is_installed.html">subread_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/stringtie_is_installed.html">stringtie_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="../reference/taco_is_installed.html">taco_is_installed</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="retrieving-sample-metadata">Retrieving sample metadata<a class="anchor" aria-label="anchor" href="#retrieving-sample-metadata"></a>
</h2>
<p>First of all, you need to choose which samples you want to download
and create a <strong>metadata data frame</strong> for your samples. To
create this data frame, you will pass a search term to the function
<code><a href="../reference/create_sample_info.html">create_sample_info()</a></code>. The search term has the same syntax
of the SRA search term syntax. For example, you can search by:</p>
<ul>
<li>BioSample accession - <strong>SAMN08903403[BSPL]</strong>
</li>
<li>BioProject accession - <strong>PRJNA229998[GPRJ]</strong>
</li>
<li>All RNA-seq samples for an organism - <strong>Glycine max[ORGN] AND
RNA-seq[STRA]</strong>
</li>
<li>And so on…</li>
</ul>
<p>Let’s create a metadata data frame for a human RNA-seq sample that is
included in the <em><a href="https://bioconductor.org/packages/3.15/airway" class="external-link">airway</a></em>
Bioconductor package.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create metadata data frame</span></span>
<span><span class="va">term</span> <span class="op">&lt;-</span> <span class="st">"SAMN02422669[BSPL]"</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_sample_info.html">create_sample_info</a></span><span class="op">(</span><span class="va">term</span><span class="op">)</span></span>
<span><span class="va">metadata</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downloading-fastq-files">Downloading FASTQ files<a class="anchor" aria-label="anchor" href="#downloading-fastq-files"></a>
</h2>
<p>To download the .fastq files with reads, you have 2 options:</p>
<ul>
<li><p>Using <code><a href="../reference/download_from_ena.html">download_from_ena()</a></code> (recommended), which
downloads reads from ENA. This function does not require any external
dependency and is much faster than using <em>SRAToolkit</em>.</p></li>
<li><p>Using <code>download_fastq()</code>, which relies on
<em>SRAToolkit</em> to download reads from NCBI’s SRA. You need to have
<em>SRAToolkit</em> installed in your machine to run this function. This
function only exists for historical reasons (i.e., some people use
<em>SRAToolkit</em> for a long time and are used to it). You should try
to avoid <em>SRAToolkit</em>, because it is slow and not so easy to set
up.</p></li>
</ul>
<p>As input, you only need to give the metadata data frame and the path
the output directory where .fastq files will be stored. For example:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download sample to temporary directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">6000</span><span class="op">)</span></span>
<span><span class="va">download</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_from_ena.html">download_from_ena</a></span><span class="op">(</span><span class="va">metadata</span>, fastqdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For running time issues, copy example FASTQ files to ds$fastqdir</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"bears"</span><span class="op">)</span>, </span>
<span>                pattern <span class="op">=</span> <span class="st">".fastq.gz"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">f</span></span>
<span><span class="va">copy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span><span class="va">f</span>, <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span><span class="op">)</span></span></code></pre></div>
<p>It’s important to check your downloaded FASTQ files for download
issues that might happen when the internet connection is not stable. The
function <code>check_downloads()</code> checks your downloaded FASTQ
files to look for one of the three issues below:</p>
<ul>
<li>
<strong>Issue 1:</strong> File was not downloaded.</li>
<li>
<strong>Issue 2:</strong> For paired-end reads, only forward or
reverse file was downloaded.</li>
<li>
<strong>Issue 3:</strong> Less reads than the expected.</li>
</ul>
<p>To look for <strong>issue 3</strong>, <code>check_downloads()</code>
needs to know the expected number of reads for each run accession. This
number is available in SRA, and it can be obtained with the function
<code>get_read_count()</code>. For example:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a data frame of run accessions and their read counts</span></span>
<span><span class="va">read_count</span> <span class="op">&lt;-</span> <span class="fu">get_read_count</span><span class="op">(</span><span class="va">metadata</span>, run_accession <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">Run</span><span class="op">)</span></span>
<span><span class="va">read_count</span></span></code></pre></div>
<p>As you can see, this function returns a data frame with run
accessions and their number of reads as reported by SRA. With this
information, you can run <code>check_downloads()</code> to look for
download problems.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check downloaded files for issues</span></span>
<span><span class="fu">check_downloads</span><span class="op">(</span><span class="va">metadata</span>, fastqdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span>, <span class="va">read_count</span><span class="op">)</span></span></code></pre></div>
<p>Here, as you can see, the function reported an issue 3 (<em>less
reads than the expected</em>). That is because we are using the example
FASTQ files that come with this package, which have been filtered for
package size issues.</p>
</div>
<div class="section level2">
<h2 id="sequence-quality-check">Sequence quality check<a class="anchor" aria-label="anchor" href="#sequence-quality-check"></a>
</h2>
<p>After downloading the .fastq files, you need to perform some quality
checks to check if you don’t have poor-quality base calling or sequence
adapters. These can be done with the functions <code><a href="../reference/run_fastqc.html">run_fastqc()</a></code>
and <code><a href="../reference/multiqc.html">multiqc()</a></code>, which will run <em>FastQC</em> <span class="citation">(Andrews et al. 2019)</span> and summarize
<em>FastQC</em> results with MultiQC <span class="citation">(Ewels et
al. 2016)</span>.</p>
<p><strong>NOTE:</strong> Both <code><a href="../reference/run_fastqc.html">run_fastqc()</a></code> and
<code><a href="../reference/multiqc.html">multiqc()</a></code> require external tools. Throughout this vignette,
we will execute functions that call external tools inside
<code><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">Herper::with_CondaEnv()</a></code> calls.</p>
<p>Here, for the purpose of demonstration, we will use example FASTQ
files stored in /extdata.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Path to MultiQC output</span></span>
<span><span class="va">multiqc_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"multiqc"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run FastQC for all .fastq files in fastqdir</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"fastqc_env"</span>,</span>
<span>    <span class="va">fastqc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_fastqc.html">run_fastqc</a></span><span class="op">(</span></span>
<span>        <span class="va">metadata</span>, </span>
<span>        fastqdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span>,</span>
<span>        fastqcdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqcdir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fastqc</span> <span class="co"># check status</span></span>
<span></span>
<span></span>
<span><span class="co"># Summarize FastQC results with MultiQC</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"multiqc_env"</span>,</span>
<span>    <span class="va">fastqc_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multiqc.html">multiqc</a></span><span class="op">(</span><span class="va">ds</span><span class="op">$</span><span class="va">fastqcdir</span>, outdir <span class="op">=</span> <span class="va">multiqc_out</span><span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fastqc_table</span> <span class="co"># check results</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-filtering">Read filtering<a class="anchor" aria-label="anchor" href="#read-filtering"></a>
</h2>
<p>Read filtering consists in 2 steps:</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/trim_reads.html">trim_reads()</a></code> - trim adapters and low-quality bases.
This function runs <em>Trimmomatic</em> <span class="citation">(Bolger,
Lohse, and Usadel 2014)</span> and saves filtered .fastq files in a
directory named <code>filtdir</code>.</li>
<li>
<code><a href="../reference/remove_rrna.html">remove_rrna()</a></code> - remove rRNA (if any) from .fastq files.
rRNA removal relies on the <em>SortMeRNA</em> <span class="citation">(Kopylova, Noé, and Touzet 2012)</span> program.</li>
</ol>
<p>Here, as you can see in <code>fastqc_table</code>, the files passed
the “Per Base Sequence Quality” check, so we don’t need to trim
anything. For this particular case, we can skip
<code><a href="../reference/trim_reads.html">trim_reads()</a></code> and only run <code><a href="../reference/remove_rrna.html">remove_rrna()</a></code>. As
the rRNA database, we will create a directory and store an example rRNA
file in it.</p>
<p>In real-world applications, your rRNA database directory should
contain all FASTA files distributed in the <a href="https://github.com/biocore/sortmerna/tree/master/data/rRNA_databases" class="external-link">SortMeRNA
GitHub repo</a>. However, if you think some of these files (e.g., 5s and
5.8s rRNA) are not a concern in your data set, you don’t need to include
them in the database.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a directory to store the rRNA db</span></span>
<span><span class="va">rrna_db_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"rrna"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">rrna_db_dir</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Copy the example 16S rRNA file to the db directory.</span></span>
<span><span class="va">rrna_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"bac_16s_subset.fa"</span>, package<span class="op">=</span><span class="st">"bears"</span><span class="op">)</span></span>
<span><span class="va">copy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">rrna_file</span>, to <span class="op">=</span> <span class="va">rrna_db_dir</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run SortMeRNA from the Conda env</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"sortmerna_env"</span>,</span>
<span>    <span class="va">rrna_removal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_rrna.html">remove_rrna</a></span><span class="op">(</span></span>
<span>        <span class="va">metadata</span>,</span>
<span>        fastqdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fastqdir</span>,</span>
<span>        filtdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">filtdir</span>,</span>
<span>        rrna_db_dir <span class="op">=</span> <span class="va">rrna_db_dir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span><span class="va">rrna_removal</span> <span class="co"># check status</span></span></code></pre></div>
<p>Now that we have performed all quality checks, we’re good to go. <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;NOTE:&lt;/strong&gt; After running
&lt;code&gt;&lt;a href="../reference/remove_rrna.html"&gt;remove_rrna()&lt;/a&gt;&lt;/code&gt;, filtered .fastq files are stored in the
directory specified in &lt;code&gt;filtdir&lt;/code&gt;. If you think this step is
not necessary for your data set and want to skip it, the
&lt;code&gt;filtdir&lt;/code&gt; directory will remain empty, so all your .fastq
files will be stored in &lt;code&gt;fastqdir&lt;/code&gt;. Keep this in mind when
specifying the path to .fastq files in the following steps of the
pipeline. If that is your case, you can make &lt;code&gt;fastqdir&lt;/code&gt; and
&lt;code&gt;filtdir&lt;/code&gt; the same with
&lt;code&gt;ds$filtdir &amp;lt;- ds$fastqdir&lt;/code&gt;.&lt;/p&gt;'><sup>1</sup></a></p>
</div>
<div class="section level2">
<h2 id="quantification-of-transcript-abundance">Quantification of transcript abundance<a class="anchor" aria-label="anchor" href="#quantification-of-transcript-abundance"></a>
</h2>
<p>Quantification of transcript abundance can be done in two ways:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Alignment-based approaches</strong>, which involves mapping
reads to a reference genome using <em>STAR</em> <span class="citation">(Dobin et al. 2013)</span> and quantifying the
expression based on uniquely mapped reads with <em>featureCounts</em>
<span class="citation">(Liao, Smyth, and Shi 2014)</span> (in raw
counts) and <em>StringTie</em> <span class="citation">(Pertea et al.
2015)</span> (in TPM).</li>
<li>
<strong>Alignment-free approaches</strong>, which involves
pseudo-aligning or quasi-mapping reads to a reference transcriptome with
<em>kallisto</em> <span class="citation">(Bray et al. 2016)</span> or
<em>salmon</em> <span class="citation">(Patro et al. 2017)</span>,
respectively.</li>
</ol>
<p>Below, we will describe how to perform both approaches.</p>
<div class="section level3">
<h3 id="alignment-based-quantification">Alignment-based quantification<a class="anchor" aria-label="anchor" href="#alignment-based-quantification"></a>
</h3>
<p>To start with, we will need to map the reads to a reference genome.
In <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em>,
reads are mapped to the reference genome with the software tool
<em>STAR</em>.</p>
<div class="section level4">
<h4 id="read-mapping">Read mapping<a class="anchor" aria-label="anchor" href="#read-mapping"></a>
</h4>
<p>Here, for the purpose of demonstration, we will map reads to a subset
of the human genome. The FASTA and GTF files corresponding to the subset
of the genome are available in /extdata.</p>
<p>Before mapping reads, we need to create a genome index. This can be
done with <code><a href="../reference/star_genome_index.html">star_genome_index()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get paths to genome subset</span></span>
<span><span class="va">genome_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh37.75_subset.fa"</span>, </span>
<span>                            package<span class="op">=</span><span class="st">"bears"</span><span class="op">)</span></span>
<span><span class="va">gff_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh37.75_subset.gtf"</span>, </span>
<span>                         package<span class="op">=</span><span class="st">"bears"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create genome index</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"star_env"</span>,</span>
<span>    <span class="va">genome_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/star_genome_index.html">star_genome_index</a></span><span class="op">(</span></span>
<span>        genome_path <span class="op">=</span> <span class="va">genome_path</span>, </span>
<span>        gff_path <span class="op">=</span> <span class="va">gff_path</span>, </span>
<span>        mappingdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span><span class="va">genome_idx</span> <span class="co"># check status</span></span></code></pre></div>
<p>Now that we have the genome index, we can map reads to it.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Map reads to the genome</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"star_env"</span>,</span>
<span>    <span class="va">read_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/star_align.html">star_align</a></span><span class="op">(</span></span>
<span>        <span class="va">metadata</span>, </span>
<span>        filtdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">filtdir</span>,</span>
<span>        fastqc_table <span class="op">=</span> <span class="va">fastqc_table</span>,</span>
<span>        mappingdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span>,</span>
<span>        gff_path <span class="op">=</span> <span class="va">gff_path</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Finally, let’s get read mapping statistics.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Obtaining read mapping statistics</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"multiqc_env"</span>,</span>
<span>    <span class="va">star_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multiqc.html">multiqc</a></span><span class="op">(</span></span>
<span>        dir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span>, </span>
<span>        outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        runon <span class="op">=</span> <span class="st">"star"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span><span class="va">star_stats</span></span></code></pre></div>
<p>Now, let’s check if samples passed the minimum quality criteria.
Here, samples are excluded if:</p>
<ol style="list-style-type: decimal">
<li>&gt;=50% of the reads fail to map or;</li>
<li>&gt;=40% of the reads fail to uniquely map.</li>
</ol>
<p>The function <code><a href="../reference/mapping_pass.html">mapping_pass()</a></code> takes the metadata data
frame and returns the same data frame, but only with the samples that
passed the minimum criteria.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check if samples passed the filtering criterion</span></span>
<span><span class="va">align_passed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapping_pass.html">mapping_pass</a></span><span class="op">(</span><span class="va">star_stats</span>, <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="va">align_passed</span> <span class="co"># inspect data</span></span>
<span></span>
<span><span class="co"># Compare to the original data set</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">align_passed</span><span class="op">)</span></span></code></pre></div>
<p>As you can see, the sample we used for read mapping passed the
minimum quality criteria. Good, huh? We can now proceed to the next
step.</p>
</div>
<div class="section level4">
<h4 id="inferring-library-strandedness">Inferring library strandedness<a class="anchor" aria-label="anchor" href="#inferring-library-strandedness"></a>
</h4>
<p>Before quantification, we need to infer library strandedness with the
<em>RSeQC</em> <span class="citation">(Wang, Wang, and Li 2012)</span>
tool. The function <code><a href="../reference/infer_strandedness.html">infer_strandedness()</a></code> runs <em>RSeQC</em>
and returns the metadata data frame with an additional column named
<strong>Orientation</strong> containing library strandedness
information.</p>
<p>This function requires the annotation in BED format, not GTF/GFF. To
convert from GTF/GFF to BED, use the function
<code><a href="../reference/gff2bed.html">gff2bed()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert GFF to BED</span></span>
<span><span class="va">bedpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gff2bed.html">gff2bed</a></span><span class="op">(</span><span class="va">gff_path</span><span class="op">)</span></span>
<span><span class="va">bedpath</span> <span class="co"># check status</span></span>
<span></span>
<span><span class="co"># Infer strandedness</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"rseqc_env"</span>,</span>
<span>    <span class="va">new_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_strandedness.html">infer_strandedness</a></span><span class="op">(</span></span>
<span>        mapping_passed <span class="op">=</span> <span class="va">align_passed</span>,</span>
<span>        bedpath <span class="op">=</span> <span class="va">bedpath</span><span class="op">$</span><span class="va">bed_path</span>,</span>
<span>        mappingdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">new_metadata</span></span></code></pre></div>
<p>As you can see, there is a new <strong>Orientation</strong> column
with strandedness info for this sample.</p>
</div>
<div class="section level4">
<h4 id="using-featurecounts">Using featureCounts<a class="anchor" aria-label="anchor" href="#using-featurecounts"></a>
</h4>
<p>Now that we have the BAM files from <em>STAR</em> and information on
library strandedness for each sample, we quantify the expression with
<em>featureCounts</em>. This tool quantifies gene expression measured in
raw read counts per gene.</p>
<p>To quantify gene expression with <em>featureCounts</em>, use the
function <code><a href="../reference/fcount.html">fcount()</a></code>. This function runs
<em>featureCounts</em> and returns a gene expression matrix with genes
in rows and samples in columns.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get gene expression in raw read counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"subread_env"</span>,</span>
<span>    <span class="va">fcounts_quant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fcount.html">fcount</a></span><span class="op">(</span></span>
<span>        <span class="va">new_metadata</span>, </span>
<span>        mappingdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span>,</span>
<span>        gff_path <span class="op">=</span> <span class="va">gff_path</span>,</span>
<span>        fcountsdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">fcountsdir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Explore gene expression matrix</span></span>
<span><span class="va">fcounts_quant</span> </span></code></pre></div>
<p>Whenever you are working with gene expression data, we recommend
storing your data as <code>SummarizedExperiment</code> objects, so you
have the expression matrix and sample metadata in a single object. If
you are not familiar with <code>SummarizedExperiment</code> objects,
take a look at the documentation of the <em><a href="https://bioconductor.org/packages/3.15/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
package.</p>
<p>To get a <code>SummarizedExperiment</code> object from
<em>featureCounts</em>, use the function
<code><a href="../reference/featureCounts2se.html">featureCounts2se()</a></code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a SummarizedExperiment object from featureCounts output</span></span>
<span><span class="va">fcountsse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/featureCounts2se.html">featureCounts2se</a></span><span class="op">(</span></span>
<span>    <span class="va">new_metadata</span>, fc_output <span class="op">=</span> <span class="va">fcounts_quant</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the SummarizedExperiment object</span></span>
<span><span class="va">fcountsse</span></span>
<span></span>
<span><span class="co"># Exploring sample metadata</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">fcountsse</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Exploring gene expression matrix</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">fcountsse</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="using-stringtie">Using StringTie<a class="anchor" aria-label="anchor" href="#using-stringtie"></a>
</h4>
<p><em>StringTie</em> quantifies transcript-level and gene-level
transcript abundances in normalized values (transcripts per million,
TPM).</p>
<p>To obtain gene expression levels in TPM with <em>StringTie</em>, use
the function <code><a href="../reference/stringtie_quantify.html">stringtie_quantify()</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Quantify expression in TPM with StringTie</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"stringtie_env"</span>,</span>
<span>    <span class="va">stringtie_quant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stringtie_quantify.html">stringtie_quantify</a></span><span class="op">(</span></span>
<span>        <span class="va">new_metadata</span>,</span>
<span>        fastqc_table <span class="op">=</span> <span class="va">fastqc_table</span>,</span>
<span>        mappingdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span>,</span>
<span>        gff_path <span class="op">=</span> <span class="va">gff_path</span>,</span>
<span>        stringtiedir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">stringtiedir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">stringtie_quant</span> <span class="co"># Check status</span></span></code></pre></div>
<p>Now, let’s read the output from <em>StringTie</em> as a
<code>SummarizedExperiment</code> object with the function
<code>stringtie2se</code>. You can choose if you want the expression at
the gene level, at the transcript level, or both. Here, let’s get the
gene-level expression. For that, you will need to give a 2-column data
frame with transcript IDs and their corresponding genes.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load transcript-to-gene correspondence</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tx2gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tx2gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read StringTie output as a SummarizedExperiment object</span></span>
<span><span class="va">stringtiese</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stringtie2se.html">stringtie2se</a></span><span class="op">(</span></span>
<span>    <span class="va">new_metadata</span>,</span>
<span>    stringtiedir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">stringtiedir</span>,</span>
<span>    level <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>    tx2gene <span class="op">=</span> <span class="va">tx2gene</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Exploring the SummarizedExperiment object</span></span>
<span><span class="va">stringtiese</span></span>
<span></span>
<span><span class="co"># Looking at gene expression matrix</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">stringtiese</span>, <span class="st">"gene_TPM"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="bonus-transcript-assembly-and-merging">Bonus: Transcript assembly and merging<a class="anchor" aria-label="anchor" href="#bonus-transcript-assembly-and-merging"></a>
</h4>
<p>Besides quantifying transcript abundance, <em>StringTie</em> can also
be used to assemble transcripts for each BioSample. Assembled
transcripts for each BioSample are represented as .gtf files.</p>
<p>However, if you want a single .gtf file with the assembled
transcripts for all BioSamples you are studying, you can merge the
individual .gtf files from <em>StringTie</em> with the software tool
<em>TACO</em> <span class="citation">(Niknafs et al. 2017)</span>.
Below, we will demonstrate how that can be achieved.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Transcript assembly with StringTie</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"stringtie_env"</span>,</span>
<span>    <span class="va">assembled_transcripts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stringtie_assemble.html">stringtie_assemble</a></span><span class="op">(</span></span>
<span>        <span class="va">new_metadata</span>,</span>
<span>        fastqc_table <span class="op">=</span> <span class="va">fastqc_table</span>,</span>
<span>        mappingdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">mappingdir</span>,</span>
<span>        gff_path <span class="op">=</span> <span class="va">gff_path</span>,</span>
<span>        stringtiedir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">stringtiedir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">assembled_transcripts</span> <span class="co"># check status</span></span></code></pre></div>
<p>In this vignette, we have a single BioSample. However, in real-life
scenarios, you would have several samples. To merge the .gtf files for
each sample in a single .gtf file, use the function
<code><a href="../reference/taco_merge.html">taco_merge()</a></code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge assembled transcripts with TACO</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"taco_env"</span>,</span>
<span>    <span class="va">merged_transcripts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taco_merge.html">taco_merge</a></span><span class="op">(</span></span>
<span>        <span class="va">new_metadata</span>,</span>
<span>        stringtiedir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">stringtiedir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged_transcripts</span> <span class="co"># check status</span></span></code></pre></div>
<p>The merged transcript assembly will be stored in a file named
<strong>final_assembly.gtf</strong> in the subdirectory
<em>assembly/merged_assembly</em>, inside <code>stringtiedir</code>. To
get the path to the .gtf file, use:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get path to merged transcript assembly</span></span>
<span><span class="va">final_assembly</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span></span>
<span>    <span class="va">ds</span><span class="op">$</span><span class="va">stringtiedir</span>, <span class="st">"assembly"</span>, <span class="st">"merged_assembly"</span>, <span class="st">"final_assembly.gtf"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">final_assembly</span></span></code></pre></div>
<blockquote>
<p>But why would I want to assemble transcripts and merge them if I
already have a .gtf file the transcript annotations?</p>
</blockquote>
<p>That’s a great question! This is a way to identify novel transcripts
that are not present in your reference .gtf file. Some transcripts can
be missing in the reference annotation (.gtf file) mainly because: i.
genome assembly does not have a good quality, so these transcripts could
not be predicted. ii. false-positives from the transcript annotation
software tool that was used.</p>
<p>If you want to have a more comprehensive transcript abundance
quantification, you can assemble transcripts for each sample, merge
them, and input the output file <strong>final_assembly.gtf</strong> to
the quantification functions. This way, instead of using the reference
transcript annotation, you will use your own transcript annotation,
which may contain novel transcripts.</p>
</div>
</div>
<div class="section level3">
<h3 id="alignment-free-quantification">Alignment-free quantification<a class="anchor" aria-label="anchor" href="#alignment-free-quantification"></a>
</h3>
<p>To quantify the expression without mapping reads to the genome, you
have two options:</p>
<ul>
<li>
<strong>Using kallisto</strong>, which quantifies transcript
abundance based on pseudoalignments.</li>
<li>
<strong>Using salmon</strong>, which quantifies transcript abundance
based on quasi-mapping.</li>
</ul>
<p>For both kallisto and salmon, you will need to have a
<strong>reference transcriptome</strong>, not a reference genome. This
is a FASTA file containing the sequences of all annotated transcripts in
your genome. You can easily create this file with the function
<code>extractTranscriptSeqs()</code> from the <em><a href="https://bioconductor.org/packages/3.15/GenomicFeatures" class="external-link">GenomicFeatures</a></em>
package.</p>
<div class="section level4">
<h4 id="using-salmon">Using salmon<a class="anchor" aria-label="anchor" href="#using-salmon"></a>
</h4>
<p>First of all, we will need to index the reference transcriptome with
the function <code><a href="../reference/salmon_index.html">salmon_index()</a></code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Path to reference transcriptome</span></span>
<span><span class="va">transcriptome_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>, <span class="st">"Homo_sapiens.GRCh37.75_subset_transcripts.fa.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"bears"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Index the transcriptome</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"salmon_env"</span>,</span>
<span>    <span class="va">idx_salmon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/salmon_index.html">salmon_index</a></span><span class="op">(</span></span>
<span>        salmonindex <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">salmonindex</span>,</span>
<span>        transcriptome_path <span class="op">=</span> <span class="va">transcriptome_path</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">idx_salmon</span> <span class="co"># check status</span></span></code></pre></div>
<p>Now, we can quantify transcript abundance with
<code><a href="../reference/salmon_quantify.html">salmon_quantify()</a></code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Quantify transcript abundance</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"salmon_env"</span>,</span>
<span>    <span class="va">quant_salmon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/salmon_quantify.html">salmon_quantify</a></span><span class="op">(</span></span>
<span>        <span class="va">new_metadata</span>,</span>
<span>        filtdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">filtdir</span>,</span>
<span>        salmonindex <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">salmonindex</span>,</span>
<span>        salmondir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">salmondir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">quant_salmon</span> <span class="co"># check status</span></span></code></pre></div>
<p>After running <code><a href="../reference/salmon_quantify.html">salmon_quantify()</a></code>, salmon output in .sf
format will be stored in the directory specified in
<code>salmondir</code>.</p>
<p>To read salmon output as a <code>SummarizedExperiment</code> object,
use the function <code><a href="../reference/salmon2se.html">salmon2se()</a></code>. You can choose if you want
the expression at the gene level, at the transcript level, or both.
Here, let’s get the gene-level expression. For that, you will need to
give a 2-column data frame with transcript IDs and their corresponding
genes.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load transcript-to-gene data frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tx2gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tx2gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read salmon output as a SummarizedExperiment object</span></span>
<span><span class="va">salmonse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/salmon2se.html">salmon2se</a></span><span class="op">(</span></span>
<span>    <span class="va">new_metadata</span>, </span>
<span>    level <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    salmondir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">salmondir</span>,</span>
<span>    <span class="va">tx2gene</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Exploring the output</span></span>
<span><span class="va">salmonse</span></span>
<span></span>
<span><span class="co"># Get gene expression matrix in TPM</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">salmonse</span>, <span class="st">"gene_TPM"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get gene expression matrix as raw counts</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">salmonse</span>, <span class="st">"gene_counts"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="using-kallisto">Using kallisto<a class="anchor" aria-label="anchor" href="#using-kallisto"></a>
</h4>
<p>Like we do in salmon, we will start by indexing the
transcriptome.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Index the transcriptome</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"kallisto_env"</span>,</span>
<span>    <span class="va">idx_kallisto</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kallisto_index.html">kallisto_index</a></span><span class="op">(</span></span>
<span>        kallistoindex <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">kallistoindex</span>,</span>
<span>        transcriptome_path <span class="op">=</span> <span class="va">transcriptome_path</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">idx_kallisto</span> <span class="co"># check status</span></span></code></pre></div>
<p>Now, we can quantify the transcript abundance.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Quantify transcript abundance</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Herper/man/UseEnvironments.html" class="external-link">with_CondaEnv</a></span><span class="op">(</span></span>
<span>    <span class="st">"kallisto_env"</span>,</span>
<span>    <span class="va">quant_kallisto</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kallisto_quantify.html">kallisto_quantify</a></span><span class="op">(</span></span>
<span>        <span class="va">new_metadata</span>,</span>
<span>        <span class="va">fastqc_table</span>, </span>
<span>        filtdir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">filtdir</span>,</span>
<span>        kallistoindex <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">kallistoindex</span>,</span>
<span>        kallistodir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">kallistodir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    pathToMiniConda <span class="op">=</span> <span class="va">my_miniconda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">quant_kallisto</span> <span class="co"># check status</span></span></code></pre></div>
<p>To read kallisto output in a SummarizedExperiment object, use the
function <code><a href="../reference/kallisto2se.html">kallisto2se()</a></code>. Again, you will have specify if you
want the expressiona at the gene level, transcript level, or both. Let’s
get the gene expression here.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read kallisto output to SummarizedExperiment object</span></span>
<span><span class="va">kallistose</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kallisto2se.html">kallisto2se</a></span><span class="op">(</span></span>
<span>    <span class="va">new_metadata</span>, </span>
<span>    level <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>    kallistodir <span class="op">=</span> <span class="va">ds</span><span class="op">$</span><span class="va">kallistodir</span>,</span>
<span>    <span class="va">tx2gene</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Exploring the output</span></span>
<span><span class="va">kallistose</span></span>
<span></span>
<span><span class="co"># Get gene expression matrix in TPM</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">kallistose</span>, <span class="st">"gene_TPM"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get gene expression matrix as raw counts</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">kallistose</span>, <span class="st">"gene_counts"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="closing-remarks">Closing remarks<a class="anchor" aria-label="anchor" href="#closing-remarks"></a>
</h2>
<p>If you are using <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em>, there are
two things you must keep in mind. First, this package was designed to be
as complete as possible, which means you don’t need to run the complete
pipeline for your own project. For instance, if you just want gene
expression values in TPM for a particular BioProject or set of
BioProjects, you can simply go through the <em>salmon</em> path of the
pipeline, skipping the read mapping section. Likewise, if you are using
<em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em> for
your own data set and you have already cleaned the reads, you can skip
the sequence quality checks and read filtering sections. The second
thing to consider is that <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em> is a work
in progress. Bioinformatics is a fast-evolving field, and new (and
better) methods to address a particular question are developed
continuously. Hence, we aim to keep <em><a href="https://github.com/almeidasilvaf/bears" class="external-link">bears</a></em> up to date
with state-of-the-art methods.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was created under the following sections:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-andrews2019" class="csl-entry">
Andrews, Simon, Felix Krueger, Anne Segonds-Pichon, Laura Biggins,
Christel Krueger, and Steven Wingett. 2019.
<span>“<span>FastQC</span>.”</span> <span>Babraham, UK</span>: Babraham
Institute.
</div>
<div id="ref-bolger2014trimmomatic" class="csl-entry">
Bolger, Anthony M, Marc Lohse, and Bjoern Usadel. 2014.
<span>“Trimmomatic: A Flexible Trimmer for Illumina Sequence
Data.”</span> <em>Bioinformatics</em> 30 (15): 2114–20.
</div>
<div id="ref-bray2016near" class="csl-entry">
Bray, Nicolas L, Harold Pimentel, Páll Melsted, and Lior Pachter. 2016.
<span>“Near-Optimal Probabilistic RNA-Seq Quantification.”</span>
<em>Nature Biotechnology</em> 34 (5): 525–27.
</div>
<div id="ref-dobin2013star" class="csl-entry">
Dobin, Alexander, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris
Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R
Gingeras. 2013. <span>“STAR: Ultrafast Universal RNA-Seq
Aligner.”</span> <em>Bioinformatics</em> 29 (1): 15–21.
</div>
<div id="ref-ewels2016multiqc" class="csl-entry">
Ewels, Philip, Måns Magnusson, Sverker Lundin, and Max Käller. 2016.
<span>“MultiQC: Summarize Analysis Results for Multiple Tools and
Samples in a Single Report.”</span> <em>Bioinformatics</em> 32 (19):
3047–48.
</div>
<div id="ref-kopylova2012sortmerna" class="csl-entry">
Kopylova, Evguenia, Laurent Noé, and Hélène Touzet. 2012.
<span>“SortMeRNA: Fast and Accurate Filtering of Ribosomal RNAs in
Metatranscriptomic Data.”</span> <em>Bioinformatics</em> 28 (24):
3211–17.
</div>
<div id="ref-liao2014featurecounts" class="csl-entry">
Liao, Yang, Gordon K Smyth, and Wei Shi. 2014. <span>“featureCounts: An
Efficient General Purpose Program for Assigning Sequence Reads to
Genomic Features.”</span> <em>Bioinformatics</em> 30 (7): 923–30.
</div>
<div id="ref-niknafs2017taco" class="csl-entry">
Niknafs, Yashar S, Balaji Pandian, Hariharan K Iyer, Arul M Chinnaiyan,
and Matthew K Iyer. 2017. <span>“TACO Produces Robust Multisample
Transcriptome Assemblies from RNA-Seq.”</span> <em>Nature Methods</em>
14 (1): 68–70.
</div>
<div id="ref-patro2017salmon" class="csl-entry">
Patro, Rob, Geet Duggal, Michael I Love, Rafael A Irizarry, and Carl
Kingsford. 2017. <span>“Salmon Provides Fast and Bias-Aware
Quantification of Transcript Expression.”</span> <em>Nature Methods</em>
14 (4): 417–19.
</div>
<div id="ref-pertea2015stringtie" class="csl-entry">
Pertea, Mihaela, Geo M Pertea, Corina M Antonescu, Tsung-Cheng Chang,
Joshua T Mendell, and Steven L Salzberg. 2015. <span>“StringTie Enables
Improved Reconstruction of a Transcriptome from RNA-Seq Reads.”</span>
<em>Nature Biotechnology</em> 33 (3): 290–95.
</div>
<div id="ref-wang2012rseqc" class="csl-entry">
Wang, Liguo, Shengqin Wang, and Wei Li. 2012. <span>“RSeQC: Quality
Control of RNA-Seq Experiments.”</span> <em>Bioinformatics</em> 28 (16):
2184–85.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Fabrício Almeida-Silva, Thiago Venancio.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
